### PR TITLE
[Spark] New Authz Plan Serde Layer

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

This PR redesigned the authorization part of the spark authz module with a New Authz Plan Serde Layer.

#### Motivation

- add a general layer to describe a command, so that we can add a new command or users can add a third-party command easily according to the specification.
- get rid of the spark version check. The built-in spark commands frequently vary from version to version, which makes us hard to maintain at compile& runtime phase, and the third-party commands are hard to check by spark versions.

#### Data structure

image

Overall, we introduce 2 general basic data structures:
- CommandSpec: used to describe a command
  - `classname` as key for the read-side to get the spec by a particular command
  - pre-defined operation type
  - descriptors  
- Descriptor: used to describe an object, such as table, db, query,
  - `fieldName`: the object to get
  - `fieldExtractor`: the method to get the object; use SPI to load
  - sub-descriptors: such as columns in a table
  - etc.

#### SPI

- Extractor: implementations for `fieldExtractor`
  - key: the name of the extractor for the read-side to get itself
  - func: converting the field value to specific and general objects

#### Code Path

- Write code path
  - automatically generated default json configuration files
  - custom json configuration files for thrid-party commands
- Read code path
  - Load json as maps
  - RuleAuthorization -> PrivilegeBuilder.build -> get command desc from maps -> build privileges with the retrieved desc.

#### TODOs

- Add back the ArcticCommand
- Add delta command
- Add ways for loading custom json configuration files
- Add hudi commands
- etc

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json [hunk 1]
```diff
@@ -0,0 +1,1135 @@
...
+    "isInput" : false,
+    "setCurrentDatabaseIfMissing" : false
+  } ],
+  "operationType" : "QUERY",
+  "queryDescs" : [ {
+    "fieldName" : "sourceTable",
+    "fieldExtractor" : "LogicalPlanQueryExtractor"
+  } ]
+}, {
+  "classname" : "org.apache.spark.sql.catalyst.plans.logical.MergeIntoIcebergTable",
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json:887]: If Iceberg's command put in `table_command_spec.json` used by `PrivilegesBuilder`, how to extend this list to separate Iceberg support in a single plugin or cover more commands ?

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json:900]: This PR does not cover this case completely, `table_command_spec_custom.json` maybe used later to support customize third-party commands

**Comment by yaooqinn** [comment]
cc @bowenliang123 @ulysses-you @pan3793 PTAL when you have time

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Descriptor.scala [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    fieldExtractor: String,
+    databaseDesc: Option[DatabaseDesc] = None,
+    functionTypeDesc: Option[FunctionTypeDesc] = None,
+    isInput: Boolean = false) extends Descriptor {
+  override def getValue(v: AnyRef): Function = {
+    val functionVal = invoke(v, fieldName)
+    val functionExtractor = functionExtractors(fieldExtractor)
+    var function = functionExtractor(functionVal)
+    if (function.database.isEmpty) {
+      function = function.copy(database = databaseDesc.map(_.getValue(v)))
```
Comment (ulysses-you) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Descriptor.scala:119]: why `TableDesc. getValue` does not need fill database ?

Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Descriptor.scala:119]: I think `TableDesc.getValue` is filling database of `Table` in `TableExtractor` implementations.
Is this a question for FunctionDesc or else?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Descriptor.scala [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+import org.apache.kyuubi.plugin.spark.authz.serde.FunctionType.FunctionType
+import org.apache.kyuubi.plugin.spark.authz.serde.FunctionTypeExtractor.functionTypeExtractors
+import org.apache.kyuubi.plugin.spark.authz.serde.QueryExtractor.queryExtractors
+import org.apache.kyuubi.plugin.spark.authz.serde.TableExtractor.tableExtractors
+import org.apache.kyuubi.plugin.spark.authz.serde.TableType.TableType
+import org.apache.kyuubi.plugin.spark.authz.serde.TableTypeExtractor.tableTypeExtractors
+import org.apache.kyuubi.plugin.spark.authz.util.AuthZUtils._
+
+/**
+ * A database object(such as database, table, function) descriptor describes its name and getter
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Descriptor.scala:39]: these are literally database objects

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/JsonSpecFileGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  def write(data: Any, filename: String): Unit = {
+    val pluginHome = getClass.getProtectionDomain.getCodeSource.getLocation.getPath
+      .split("target").head
+    val writer = {
+      val p =
+        Paths.get(pluginHome, "src", "main", "resources", s"${filename}_command_spec.json")
+      Files.newBufferedWriter(p, StandardCharsets.UTF_8)
+    }
+    mapper.writerWithDefaultPrettyPrinter().writeValue(writer, data)
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/JsonSpecFileGenerator.scala:47]: Consider sort `data` array by class name before writting to spec json, for readability and making commands in group order.
`.sortBy(_.classname)`

**Comment by bowenliang123** [comment]
+1, LGTM. And thanks for the effort. Impressively passed the all the existed auth test in CI with in several commits.

It's a great work for redesigning and modernizing Authz core implementation. It significantly improves the pattern and clarification in rule authentication by components in layers of specs, descriptors, and extractors.

The idea and implemented components in resource privilege checking can be also applied to row filtering, column masking, object filtering, and even to other fields like lineage.

**Review by bowenliang123 — APPROVED** [review]

**Comment by bowenliang123** [comment]
fyi @jeanlyn

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/JsonSpecFileGenerator.scala (hunk 4 above)
Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/JsonSpecFileGenerator.scala]: addressed

**Comment by ulysses-you** [comment]
late lgtm

**Pull request closed without merging**